### PR TITLE
[Bugfix][MRV2][LoRA] Fix prompt logprobs mapping

### PR DESCRIPTION
**Pull request by HuskyLYL**
## Purpose

Model Runner V2 splits prompt logprobs computation into smaller chunks to reduce GPU memory usage. However, the lm_head LoRA metadata was not rebuilt for each chunk and continued to use the full-batch mapping.

This PR rebuilds the token-level LoRA metadata for every prompt-logprobs chunk, ensuring that the correct LoRA mapping is applied and preventing incorrect logits or CUDA memory access errors.

## Test Plan

Tested Qwen3.5-4B with Model Runner V2 on one NVIDIA GPU using eight distinct `lm_head` LoRA adapters. We sent 32 concurrent requests with `logprobs=5` and `prompt_logprobs=5`, using 1025 total prompt tokens. With vLLM's default 1024-token chunk size, the prompt was split into `1024 + 1` chunks. Serial and concurrent results were compared to verify LoRA mapping. 

This PR extends the V1 fix from #51597 to Model Runner V2.

## Test Result

On unpatched `main` at `d9fbe526c0`, the average logprob difference was `0.52–0.55`, with a maximum difference of `25.3`. 

On `fix/mrv2-lora-prompt-logprobs-mapping` at `b4964b1bd8`, the average difference decreased to `0.059–0.063`, with a maximum below `0.92`, close to the no-LoRA baseline. No NaN values, CUDA illegal memory accesses, EngineCore crashes, or request failures were observed.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Review by linitra24 — COMMENTED** [review]
Overall, LGTM. One question though: why do we need to add a new helper function here instead of reusing `update_metadata` to update the logits metadata?

**Comment by HuskyLYL** [comment]
> Overall, LGTM. One question though: why do we need to add a new helper function here instead of reusing `update_metadata` to update the logits metadata?

Thanks for the review and the question, @linitra24.

`update_metadata` prepares metadata for the original MRV2 forward batch, including the full-batch token mapping. MRV2 later splits prompt hidden states into 1024-token chunks to reduce logits memory usage, so `update_metadata()` does not prepare separate metadata for these chunk-level logits computations.

**Comment by HuskyLYL** [comment]
> Overall, LGTM. One question though: why do we need to add a new helper function here instead of reusing `update_metadata` to update the logits metadata?

In practice, we use the full token-level mapping prepared by `update_metadata()` and slice it for each chunk.

**Comment by linitra24** [comment]
Thanks for the explanation. However, I still think the implementation in https://github.com/vllm-project/vllm/pull/51597 looks a bit simpler. Also, it seems that the current PR does not address the similar issue in the V1 model runner.

**Comment by HuskyLYL** [comment]
> Thanks for the explanation. However, I still think the implementation in #51597 looks a bit simpler. Also, it seems that the current PR does not address the similar issue in the V1 model runner.

Thanks for the helpful suggestion and feedback.

After taking another look, I agree that the current implementation is more defensive than necessary when restoring the metadata. Since each chunk overwrites the logits mapping before the next `compute_logits()` call, we only need to restore the original mapping once after the chunk loop.

This allows us to remove the extra helper and scratch-state management and simplify the MRV2 implementation following the approach in #51597. I will start refactoring it accordingly.

**Comment by HuskyLYL** [comment]
> Thanks for the explanation. However, I still think the implementation in #51597 looks a bit simpler. Also, it seems that the current PR does not address the similar issue in the V1 model runner.

@linitra24 Yes, the V1 code path is intentionally not included to avoid duplicating work. We can address the same issue in the V1 model runner separately.

**Comment by linitra24** [comment]
> > Thanks for the explanation. However, I still think the implementation in #51597 looks a bit simpler. Also, it seems that the current PR does not address the similar issue in the V1 model runner.
> 
> Thanks for the helpful suggestion and feedback.
> 
> After taking another look, I agree that the current implementation is more defensive than necessary when restoring the metadata. Since each chunk overwrites the logits mapping before the next `compute_logits()` call, we only need to restore the original mapping once after the chunk loop.
> 
> This allows us to remove the extra helper and scratch-state management and simplify the MRV2 implementation following the approach in #51597. I will start refactoring it accordingly.

Thanks for the clarification. That sounds good to me. I'll take another look after the refactor!

**Comment by HuskyLYL** [comment]
Thanks for the feedback, @linitra24. I’ve completed the implementation and tested the MRV2 GPU LoRA prompt-logprobs path, including prompts crossing the 1024-token chunk boundary.

Please feel free to suggest any further changes or improvements.

File: vllm/v1/worker/gpu/sample/prompt_logprob.py [hunk 1]
```diff
@@ -211,24 +224,42 @@ def compute_prompt_logprobs_with_chunking(
...
+        getattr(lora_wrapper, "prompt_mapping_meta", None)
+        if lora_wrapper is not None
+        else None
+    )
+    try:
+        for start_idx in range(0, prompt_token_ids.shape[0], CHUNK_SIZE):
+            end_idx = start_idx + CHUNK_SIZE
+            # NOTE(woosuk): logits_fn can be slow because it involves all-gather.
+            if prompt_mapping_meta is not None:
+                assert lora_wrapper is not None
```
Comment (linitra24) [inline vllm/v1/worker/gpu/sample/prompt_logprob.py:238]: Does this assertion serve any purpose here?

File: vllm/v1/worker/gpu/sample/prompt_logprob.py [hunk 2]
```diff
@@ -211,24 +224,42 @@ def compute_prompt_logprobs_with_chunking(
...
+                requested_num,
+                prompt_token_ids[start_idx:end_idx],
+                logits_mode=logits_mode,
+            )
+            token_ids.append(result.logprob_token_ids)
+            scores.append(result.logprobs)
+            ranks.append(result.selected_token_ranks)
+    finally:
+        if prompt_mapping_meta is not None:
+            assert lora_wrapper is not None
```
Comment (linitra24) [inline vllm/v1/worker/gpu/sample/prompt_logprob.py:260]: Same as above.

File: vllm/v1/worker/gpu/sample/prompt_logprob.py [hunk 3]
```diff
@@ -211,24 +224,42 @@ def compute_prompt_logprobs_with_chunking(
...
+        if lora_wrapper is not None
+        else None
+    )
+    try:
+        for start_idx in range(0, prompt_token_ids.shape[0], CHUNK_SIZE):
+            end_idx = start_idx + CHUNK_SIZE
+            # NOTE(woosuk): logits_fn can be slow because it involves all-gather.
+            if prompt_mapping_meta is not None:
+                assert lora_wrapper is not None
+                with gpu_sync_allowed():
```
Comment (linitra24) [inline vllm/v1/worker/gpu/sample/prompt_logprob.py:239]: prepare_tensors() introduces a GPU-to-CPU synchronization for every 1024-token chunk. Would this have a noticeable impact on performance?

Comment (HuskyLYL) [reply vllm/v1/worker/gpu/sample/prompt_logprob.py:239]: I tested the MRV2 prompt-logprobs path on an NVIDIA A800  with TP1,using Qwen3.5-4B and one rank-4 LoRA adapter. I compared preparing LoRA metadata for every 1024-token chunk with preparing it only once. Prompt lengths ranged from 1K to 64K tokens, with 10 warmup runs and 5 measured runs.

I measured the full `PromptLogprobsWorker.compute_prompt_logprobs()` call,including token processing, `prepare_tensors()`, logits computation, top-klogprob calculation, and result assembly.

| Prompt length | Prepare per chunk | Prepare once | Additional overhead |
|---:|---:|---:|---:|
| 1K | 10.918 ms | 10.981 ms | -0.063 ms |
| 2K | 21.331 ms | 20.829 ms | 0.502 ms |
| 4K | 41.746 ms | 40.701 ms | 1.045 ms |
| 8K | 83.594 ms | 80.645 ms | 2.949 ms |
| 16K | 166.071 ms | 159.949 ms | 6.122 ms |
| 32K | 329.405 ms | 317.720 ms | 11.684 ms |
| 64K | 656.167 ms | 631.425 ms | 24.741 ms |

The additional cost is approximately 0.4 ms per extra 1024-token chunk. At 64K tokens, preparing metadata for every chunk adds about 24.7 ms, or 3.8% of the total `PromptLogprobsWorker` time.

Comment (HuskyLYL) [reply vllm/v1/worker/gpu/sample/prompt_logprob.py:239]: For a 64K prompt, repeated metadata preparation adds approximately 3.8% overhead within `PromptLogprobsWorker`. Relative to the complete vLLM E2E time from inference start to prompt logprobs generation, it accounts for approximately 0.45%.

File: vllm/v1/worker/gpu/sample/prompt_logprob.py (hunk 1 above)
Comment (HuskyLYL) [reply vllm/v1/worker/gpu/sample/prompt_logprob.py:238]: > Does this assertion serve any purpose here?

Yes. The assertion is needed for mypy to narrow `lora_wrapper` from an optional value. Removing it causes a `union-attr` error in this file.

File: vllm/v1/worker/gpu/sample/prompt_logprob.py (hunk 3 above)
Comment (HuskyLYL) [reply vllm/v1/worker/gpu/sample/prompt_logprob.py:239]: One possible approach would be to scan the complete CPU-side token mapping once before entering the chunk loop and identify which chunks require a metadata update. During the loop, we would call `prepare_tensors()` only when the mapping changes and reuse the existing metadata for the other chunks.
